awsbestpractices/loops-to-batches/LoopsToBatches.java

```diff
@@ -0,0 +1,185 @@
+    void wrapper() {
+        for (Record record : records) {
+            // Wrapper/proxy - should not flag it.
+            StreamsAdapterDemoHelper.putItem(dynamoDBClient, tableName, streamRecord.getDynamodb().getNewImage());
+        }
+    }
+
+    // Based on https://github.com/travel-cloud/Cheddar/blob/43d8d3f6c26459ca9e3347682a2bcac2a79c0560/cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/AbstractDynamoDbTemplate.java#L158
+    protected final <T extends Item> Collection<PropertyDescriptor> batchablePutInLongComplexLoop(final T item,
+                                                                                                  final ItemConfiguration itemConfiguration,
+                                                                                                  final Collection<PropertyDescriptor> constraintPropertyDescriptors) {
+        final Set<PropertyDescriptor> createdConstraintPropertyDescriptors = new HashSet<>();
+        ItemConstraintViolationException itemConstraintViolationException = null;
+        for (final UniqueConstraint uniqueConstraint : itemConfiguration.uniqueConstraints()) {
+            final String uniqueConstraintPropertyName = uniqueConstraint.propertyName();
+            final PropertyDescriptor uniqueConstraintPropertyDescriptor = uniqueConstraint.propertyDescriptor();
+            if (constraintPropertyDescriptors.contains(uniqueConstraintPropertyDescriptor)) {
+                final AttributeValue uniqueConstraintAttributeValue = DynamoDbPropertyMarshaller.getValue(item,
+                        uniqueConstraintPropertyDescriptor);
+                if (uniqueConstraintAttributeValue == null) {
+                    continue;
+                }
+                if (uniqueConstraintAttributeValue.getS() != null) {
+                    uniqueConstraintAttributeValue.setS(uniqueConstraintAttributeValue.getS().toUpperCase());
+                }
+                final Map<String, AttributeValue> attributeMap = new HashMap<>();
+                attributeMap.put("property", new AttributeValue(uniqueConstraintPropertyName));
+                attributeMap.put("value", uniqueConstraintAttributeValue);
+                final Map<String, ExpectedAttributeValue> expectedResults = new HashMap<>();
+                expectedResults.put("value", new ExpectedAttributeValue(false));
+                final String indexTableName = databaseSchemaHolder.schemaName() + "-indexes."
+                        + itemConfiguration.tableName();
+                final PutItemRequest itemRequest = new PutItemRequest().withTableName(indexTableName)
+                        .withItem(attributeMap).withExpected(expectedResults);
+                try {
+                    dynamoDBClient.putItem(itemRequest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code is using a single API operation, but a batch API operation exists. You could help improve efficiency by using the batch operation instead. The batch operation is batchWriteItem.